[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48107 [quant][refactor] Move some util functions from torch/quantization/fx/utils.py to torch/quantization/utils.py
* #48073 [quant][fix][ez] Fix quant_type classification for fp16, fp16
* #48069 [quant][fix] Fix quant type classification for float_qparam qconfig
* #48067 [quant][graphmode][fx] Keep linear op unchanged when qconfig is not supported
* **#48062 [qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig**

Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25007537](https://our.internmc.facebook.com/intern/diff/D25007537)

jerryzh168 committed:
[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig

Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig"


Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig"


Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25007537](https://our.internmc.facebook.com/intern/diff/D25007537)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/pattern_utils.py:
```diff
@@ -37,6 +37,18 @@ def get_default_quant_patterns():
+# a set of QuantizeHandler classes that are not observed
+# we'll skip inserting observers for input and output for these QuantizeHandlers
+# used for ops that only supports dynamic/weight only quantization
+DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER = set()
+def mark_input_output_not_observed():
+    def insert(fn):
+        DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER.add(fn)
+        return fn
+    return insert
```
question (not a suggestion that we do that, more asking for the sake of discussion): what are people's thoughts about defining this in terms of the implementation (current code), versus the use case (a mapping of ops to the quantization modes supported by those ops)?

jerryzh168 replied:
current code is trying to mark the ops that only works for dynamic/weight_only quant to bypass the insert observer for these ops, so we don't really need to complete list. Is the suggestions to add a complete specification for all ops? what is the motivation for that? If this is trying for documentation purpose I think we probably need to list them explicitly, here we are talking about QuantizeHandler, which is an implementation detail of fx graph mode quantization, not talking about specific module/ops actually.

vkuzo replied:
my question is about the conceptual level of how the mapping is defined, where the choices are (a) defined by concepts internal to the implementation, versus (b) defined by concepts used by the end users (which op supports which quantization type).  Currently this code is doing (a).  A risk with (a) is it requires an understanding of the implementation to modify it, so it will be harder for people who don't work on FX graph mode full time to make changes correctly.  If defined in terms of (b), that risk would be reduced.

In terms of implementation of (b), there are ways to do it with allowlist (specify all mappings) or blocklists (specify all the exceptions), I agree that using blocklists might be nicer.

It doesn't have to be in this PR, but something to consider for the future.

jerryzh168 replied:
Sure. I don't expect these cases to come up very often actually, we can do that if that becomes common in the future. Although I don't really feel that would help too much, since in that case we also need to define another mapping from ops/modules to QuantizeHandler.

supriyar commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -454,8 +457,11 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
         dtypes = get_qconfig_dtypes(qconfig)
-        assert dtypes in supported_dtypes, "qconfig dtype pair not supported:" \
-            " {}, supported dtypes are: {}".format(dtypes, supported_dtypes)
+        if dtypes not in supported_dtypes:
```
Could you also update the comments on line 448-449 to update compute_type to float32?

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig"


Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25007537](https://our.internmc.facebook.com/intern/diff/D25007537)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -454,8 +457,11 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
-        assert dtypes in supported_dtypes, "qconfig dtype pair not supported:" \
-            " {}, supported dtypes are: {}".format(dtypes, supported_dtypes)
+        if dtypes not in supported_dtypes:
+            warnings.warn("dtype combination: {} is not "
```
Can we also include the types currently supported in this warning message?

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig"


Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25007537](https://our.internmc.facebook.com/intern/diff/D25007537)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Embedding/EmbeddingBag works in static quant qconfig"


Summary:
When Embedding/EmbeddingBag are configured with static quant we'll skip inserting observer for
them in the graph and keep the op unchanged and print a warning.
This also aligns with eager mode behavior as well.

We'll enforce this behavior for other ops that only supports dynamic/weight_only quant but not static quant as well.

We used a global variable `DEFAULT_NOT_OBSERVED_QUANTIZE_HANDLER`, this is not exposed to user right now,
we can add that later if needed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25007537](https://our.internmc.facebook.com/intern/diff/D25007537)

[ghstack-poisoned]
